docs/apache.md

```diff
@@ -7,6 +7,7 @@ Here are configuration examples for setting up apache as reverse proxy for gotif
 
```

Comment:
Can you explain, how I can reproduce the error?
I use the following config https://gist.github.com/jmattheis/1647df1393095c094ebea1a537d0a323#file-httpd-conf-L408
with 
```
docker run --rm --net=host -v $PWD/httpd.conf:/usr/local/apache2/conf/httpd.conf httpd:2.4
```
And the websocket connections appears to be working fine.

Reply:
I run docker on a host Ubuntu Server 18.04 with Apache 2.4.43, in this case apache is on a host machine, not in a docker.
I have SSL encryption enabled on a front end (Apache2) and proxying requests from a host machine to the container without encryption.

Somehow without those rules I see a web socket error in the web and in a console I can see that packages are ended with 400 as described in a https://github.com/gotify/server/pull/139.
```
10.10.0.32 - - [07/Jun/2020:15:23:07 +0200] "GET /gotify/stream?token=TOcKENxxxxx HTTP/1.1" 400 1698 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:77.0) Gecko/20100101 Firefox/77.0"
```
You can see 101 when I enable rewrite rule:

![image](https://user-images.githubusercontent.com/6813635/83969985-4e184280-a8d3-11ea-8648-a8b83ea89c7d.png)

It could be that if you run Apache in a docker, then you will not see this error, but if you forward requests from a host machine to the docker it does not work as expected. Quick googling gave me answer with mod_rewrite from stackoverflow.

Reply:
Could you append your full apache config. The one that doesn't work and the one that works?

Reply:
Sure,
This is 100-domain.conf
```
<VirtualHost 192.168.0.9:80>

	ServerName DOMAIN
	ServerAdmin webmaster@localhost

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined
    
	Redirect permanent / https://DOMAIN/

</VirtualHost>

<VirtualHost 192.168.0.9:443>

	ServerName DOMAIN
	ServerAdmin webmaster@DOMAIN

	DocumentRoot /var/www/nextcloud
	Alias /javascript /usr/share/javascript/

	#Error handling
	Alias /error /var/www/errors/
	ErrorDocument 403 /error/403.html
	ErrorDocument 404 /error/404.html
	ErrorDocument 500 /error/500.html
	ErrorDocument 501 /error/500.html
	ErrorDocument 502 /error/500.html
	ErrorDocument 503 /error/500.html
	ErrorDocument 504 /error/500.html

	#Logs
	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	SSLEngine on
		
	<FilesMatch "\.(cgi|shtml|phtml|php)$">
		SSLOptions +StdEnvVars
	</FilesMatch>

	<Directory /usr/lib/cgi-bin>
		SSLOptions +StdEnvVars
	</Directory>
        
	<Directory /var/www/nextcloud/>
		Require all granted
		Options FollowSymlinks MultiViews
#		Options +FollowSymlinks
		AllowOverride All

		<IfModule mod_dav.c>
			Dav off
		</IfModule>

		SetEnv HOME /var/www/nextcloud
		SetEnv HTTP_HOME /var/www/nextcloud
	</Directory>

	Protocols h2 h2c http/1.1

	H2Push on
	H2PushPriority * after
	H2PushPriority text/css before
	H2PushPriority image/jpg after 32
	H2PushPriority image/jpeg after 32
	H2PushPriority image/png after 32
	H2PushPriority application/javascript interleaved

	Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains; preload"
	Header always set Public-Key-Pins "pin-sha256=\"Vq4G0+fDfswxGN4/GiJZfOT4mhQWWpAwnphhqFUMCxM=\"; pin-sha256=\"sRHdihwgkaib1P1gxX8HFszlD+7/gTfNvuAybgLPNis=\"; pin-sha256=\"YLh1dUR9y6Kja30RrAn7JKnbQG/uEtLMkBgFF2Fuihg=\"; pin-sha256=\"C5+lpZ7tcVwmwQIMcRtPbsQtWLABXhQzejna0wHFr8M=\"; includeSubDomains; max-age=2592000;"
	Header always set Referrer-Policy "no-referrer"
	Header always set Referrer-Policy "strict-origin"
	Header always set Referrer-Policy "no-referrer-when-downgrade"
	Header always set Referrer-Policy "strict-origin-when-cross-origin"
	Header always set X-Content-Type-Options "nosniff"
	Header always set X-XSS-Protection "1; mode=block"
	Header always set X-Robots-Tag "none"
	Header always set X-Frame-Options "SAMEORIGIN"

	Include /etc/apache2/sites-available/800-proxy.conf

	SSLCertificateFile /etc/letsencrypt/live/DOMAIN/fullchain.pem
	SSLCertificateKeyFile /etc/letsencrypt/live/DOMAIN/privkey.pem
</VirtualHost>
```
and this is proxy.conf where all services are listed (also gotify)
```
# Turn on HTTP2 for proxy
H2Direct on

ProxyRequests Off
AllowEncodedSlashes NoDecode

#Pihole
	#Pihole rewrite link from /admin to /piholeadmin
	RewriteEngine  on
	RewriteRule    "^/admin/(.*)"  "/piholeadmin/$2" [R]

#Global Proxy
	<Proxy *>
		Order allow,deny
		Allow from all
	</Proxy>
#GIT
	# Note: no trailing slash after either /git or port
	ProxyPass /git http://localhost:3000 nocanon
	ProxyPassReverse /git http://localhost:3000

#Transmission
Redirect 301 "/transmission" "/transmission/web/"
	<Location /transmission/>
		Require ip 192.168.0.0/24 127.0.0.1

		RewriteEngine  on
		RewriteRule /transmission/(.*) http://127.0.0.1:9091/transmission/$1 [P,L]

		ProxyPass http://127.0.0.1:9091/transmission/ retry=0 timeout=5
		ProxyPassReverse http://127.0.0.1:9091/transmission/
	</Location>

#Portainer
Redirect 301 "/portainer" "/portainer/"
	<Location /portainer/>
		Require ip 192.168.0.0/24 127.0.0.1
		ProxyPass http://localhost:9000/ retry=0 timeout=5
		ProxyPassReverse http://localhost:9000/
		RequestHeader set Connection ""
	</Location>

#Organizr
Redirect 301 "/orga" "/orga/"
	<Location /orga/>
		Require ip 192.168.0.0/24 127.0.0.1
		ProxyPass http://localhost:9002/ retry=0 timeout=5
		ProxyPassReverse http://localhost:9002/
	</Location>

#Gotify
Redirect 301 "/gotify" "/gotify/"
Keepalive On
	<Location /gotify/>
		Require ip 192.168.0.0/24 127.0.0.1
		# The proxy must preserve the host because gotify verifies the host with the origin
		# for WebSocket connections
		ProxyPreserveHost On
		ProxyPass http://localhost:9007/ retry=0 timeout=5
		ProxyPassReverse http://localhost:9007/

	</Location>

	RewriteEngine  on
	RewriteCond %{HTTP:Upgrade} =websocket
	RewriteRule /gotify/stream(.*) ws://localhost:9007/stream$1 [P,L]

	# Proxy web socket requests to /stream
	ProxyPass "/gotify/stream" ws://localhost:9007/stream retry=0 timeout=5
```

Reply:
Basically this works and if you comment last 3 Rewrite Lines it dos't:

```
#Gotify
Redirect 301 "/gotify" "/gotify/"
Keepalive On
	<Location /gotify/>
		Require ip 192.168.0.0/24 127.0.0.1
		# The proxy must preserve the host because gotify verifies the host with the origin
		# for WebSocket connections
		ProxyPreserveHost On
		ProxyPass http://localhost:9007/ retry=0 timeout=5
		ProxyPassReverse http://localhost:9007/

	</Location>

#	RewriteEngine  on
#	RewriteCond %{HTTP:Upgrade} =websocket
#	RewriteRule /gotify/stream(.*) ws://localhost:9007/stream$1 [P,L]

	# Proxy web socket requests to /stream
	ProxyPass "/gotify/stream" ws://localhost:9007/stream retry=0 timeout=5
```